<a href="https://colab.research.google.com/github/Tessellate-Imaging/monk_v1/blob/master/study_roadmaps/2_transfer_learning_roadmap/5_exploring_model_families/4_resnet/1.3)%20Intro%20to%20resnet50-v1%20network%20-%20mxnet%20backend.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Goals


### Train a classifier using resnet50 on natural-images dataset

### Understand what lies inside resnet50 network

# What is resnet

## Readings on resnet

  1) Points from https://towardsdatascience.com/an-overview-of-resnet-and-its-variants-5281e2f56035
    - The core idea of ResNet is introducing a so-called “identity shortcut connection” that skips one or more layers
    - The deeper model should not produce a training error higher than its shallower counterparts.
    - solves the problem of vanishing gradiens as network depth increased - https://medium.com/@anishsingh20/the-vanishing-gradient-problem-48ae7f501257
    
    
 
  2) Points from https://medium.com/@14prakash/understanding-and-implementing-architectures-of-resnet-and-resnext-for-state-of-the-art-image-cf51669e1624
    - Won 1st place in the ILSVRC 2015 classification competition with top-5 error rate of 3.57% (An ensemble model)
    - Efficiently trained networks with 100 layers and 1000 layers also.
    - Replacing VGG-16 layers in Faster R-CNN with ResNet-101. They observed a relative improvements of 28%
    

  3) Read more here
    - https://arxiv.org/abs/1512.03385
    - https://d2l.ai/chapter_convolutional-modern/resnet.html
    - https://cv-tricks.com/keras/understand-implement-resnets/
    - https://mc.ai/resnet-architecture-explained/


# Table of Contents


## [Install](#0)


## [Load experiment with resnet base architecture](#1)


## [Visualize resnet](#2)


## [Train the classifier](#3)


## [Run inference on trained classifier](#4)

<a id='0'></a>
# Install Monk

## Using pip (Recommended)

  - colab (gpu) 
      - All bakcends: `pip install -U monk-colab`
      

  - kaggle (gpu) 
      - All backends: `pip install -U monk-kaggle`
      

  - cuda 10.2	
      - All backends: `pip install -U monk-cuda102`
      - Gluon bakcned: `pip install -U monk-gluon-cuda102`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda102`
      - Keras backend: `pip install -U monk-keras-cuda102`
      

  - cuda 10.1	
      - All backend: `pip install -U monk-cuda101`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda101`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda101`
	  - Keras backend: `pip install -U monk-keras-cuda101`
      

  - cuda 10.0	
      - All backend: `pip install -U monk-cuda100`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda100`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda100`
	  - Keras backend: `pip install -U monk-keras-cuda100`
      

  - cuda 9.2	
      - All backend: `pip install -U monk-cuda92`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda92`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda92`
	  - Keras backend: `pip install -U monk-keras-cuda92`
      

  - cuda 9.0	
      - All backend: `pip install -U monk-cuda90`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda90`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda90`
	  - Keras backend: `pip install -U monk-keras-cuda90`
      

  - cpu 		
      - All backend: `pip install -U monk-cpu`
	  - Gluon bakcned: `pip install -U monk-gluon-cpu`
	  - Pytorch backend: `pip install -U monk-pytorch-cpu`
	  - Keras backend: `pip install -U monk-keras-cpu`

## Install Monk Manually (Not recommended)
 
### Step 1: Clone the library
 - git clone https://github.com/Tessellate-Imaging/monk_v1.git
 
 
 
 
### Step 2: Install requirements 
 - Linux
     - Cuda 9.0
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu90.txt`
     - Cuda 9.2
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu92.txt`
     - Cuda 10.0
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu100.txt`
     - Cuda 10.1
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu101.txt`
     - Cuda 10.2
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu102.txt`
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Linux && pip install -r requirements_cpu.txt`
 
 
  - Windows
     - Cuda 9.0 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu90.txt`
     - Cuda 9.2 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu92.txt`
     - Cuda 10.0 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu100.txt`
     - Cuda 10.1 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu101.txt`
     - Cuda 10.2 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu102.txt`
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cpu.txt`
         
         
  - Mac
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Mac && pip install -r requirements_cpu.txt`
         
         
  - Misc
      - Colab (GPU)
          - `cd monk_v1/installation/Misc && pip install -r requirements_colab.txt`
      - Kaggle (GPU)
          - `cd monk_v1/installation/Misc && pip install -r requirements_kaggle.txt`
 
 
 
### Step 3: Add to system path (Required for every terminal or kernel run)
 - `import sys`
 - `sys.path.append("monk_v1/");`

## Dataset - Natural Images Classification
    - https://www.kaggle.com/prasunroy/natural-images

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1sbQ_KaEDd7kRrTvna-4odLqxM2G0QT0Z' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1sbQ_KaEDd7kRrTvna-4odLqxM2G0QT0Z" -O natural-images.zip && rm -rf /tmp/cookies.txt

In [2]:
! unzip -qq natural-images.zip

# Imports

In [1]:
#Using mxnet-gluon backend 

# When installed using pip
from monk.gluon_prototype import prototype


# When installed manually (Uncomment the following)
#import os
#import sys
#sys.path.append("monk_v1/");
#sys.path.append("monk_v1/monk/");
#from monk.gluon_prototype import prototype

<a id='1'></a>
# Load experiment with resnet base architecture

## Creating and managing experiments
    - Provide project name
    - Provide experiment name
    - For a specific data create a single project
    - Inside each project multiple experiments can be created
    - Every experiment can be have diferent hyper-parameters attached to it

In [2]:
gtf = prototype(verbose=1);
gtf.Prototype("Project", "resnet-intro");

Mxnet Version: 1.5.1

Experiment Details
    Project: Project
    Experiment: resnet-intro
    Dir: /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/2_transfer_learning_roadmap/5_exploring_model_families/4_resnet/workspace/Project/resnet-intro/



### This creates files and directories as per the following structure
    
    
    workspace
        |
        |--------Project
                        |
                        |
                        |-----resnet-intro
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)

## Set dataset and select the model

## Quick mode training

    - Using Default Function
        - dataset_path
        - model_name
        - freeze_base_network
        - num_epochs
        
        
## Sample Dataset folder structure

    natural-images
        |
        |-----train
                |------aeroplane
                        |
                        |------img1.jpg
                        |------img2.jpg
                        |------.... (and so on)
                |------car
                        |
                        |------img1.jpg
                        |------img2.jpg
                        |------.... (and so on)   
                |------.... (and so on)
         |
         |
         |-----val
                |------aeroplane
                        |
                        |------img1.jpg
                        |------img2.jpg
                        |------.... (and so on)
                |------car
                        |
                        |------img1.jpg
                        |------img2.jpg
                        |------.... (and so on)   
                |------.... (and so on)

In [3]:
gtf.Default(dataset_path="natural-images/train", 
            model_name="resnet50_v1", 
            freeze_base_network=False,
            num_epochs=5);

Dataset Details
    Train path:     natural-images/train
    Val path:       None
    CSV train path: None
    CSV val path:   None
    Label Type:     single

Dataset Params
    Input Size:   224
    Batch Size:   4
    Data Shuffle: True
    Processors:   8
    Train-val split:   0.7

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 4829
    Num val images:   2070
    Num classes:      8

Model Params
    Model name:           resnet50_v1
    Use Gpu:              True
    Use pretrained:       True
    Freeze base network:  False

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet50_v1
        Num of potentially trainable layers:  107

/home/ubuntu/.virtualenvs/monk_pip_test/lib/python3.6/site-packages/monk/system/imports.py:160: UserWarning: ArgumentWarning: clipnorm and clipvalue are active only for keras in current version of Monk
  warnings.warn(msg)
/home/ubuntu/.virtualenvs/monk_pip_test/lib/python3.6/site-packages/monk/system/imports.py:160: UserWarning: ArgumentWarning: momentum_dampening_rate is active only for pytorch in current version of Monk
  warnings.warn(msg)


## From the summary above

    - Model Params
        Model name:                           resnet50_v1
        Num of potentially trainable layers:  107
        Num of actual trainable layers:       107

<a id='2'></a>
# Visualize resnet

In [ ]:
gtf.Visualize_With_Netron(data_shape=(3, 224, 224), port=8082);

## resnet block - 1
  
  - Creating network and blocks using monk from scratch will be dealt in different roadmap series

In [ ]:
from IPython.display import Image
Image(filename='imgs/resnet50_v1_bottleneck_block1_mxnet.png') 

## Properties

  - This block has 2 branches
    - First branch is the identity branch, it takes the input and pushes it as the output, the Residual
    - Second branch has these layers
        - conv_1x1 -> batchnorm -> relu -> conv_3x3 -> batchnorm -> relu -> conv_1x1 -> batchnorm
        
  - The branches are added elementwise, so both the branches need to have same sized output
  
  - The final layer to this block is relu
  
  
### Unlike resnet18-v1 Block 1 this has a bottleneck block
   
   - The bottleneck
     - The num features in first and middle convolutions is input_features/4
     - The final convolution has features = input_features
     
   - Eg.
          - conv_1x1 -> batchnorm -> relu -> conv_3x3 -> batchnorm -> relu -> conv_1x1 -> batchnorm

    (feat-in)-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------(feat-out)    
       256 ----------> 64------------------------>-----------------------------64------------------------>-----------------------------256-------------------------------------------->  256    

## resnet block - 2
  
  - Creating network and blocks using monk from scratch will be dealt in different roadmap series

In [ ]:
from IPython.display import Image
Image(filename='imgs/resnet50_v1_bottleneck_block2_mxnet.png') 

## Properties

  - This block has 2 branches
    - First branch has these layers
        - conv_1x1 -> batchnorm
    - Second branch has these layers
        - conv_1x1 -> batchnorm -> relu -> conv_3x3 -> batchnorm -> relu -> conv_1x1 -> batchnorm
        
  - The branches are added elementwise, so both the branches need to have same sized output
  
  - The final layer to this block is relu
  
  
### Unlike resnet18-v1 Block 1 this has a bottleneck block
   
   - The bottleneck
     - The num features in first and middle convolutions is input_features/4
     - The final convolution has features = input_features
     
   - Eg.
          - conv_1x1 -> batchnorm -> relu -> conv_3x3 -> batchnorm -> relu -> conv_1x1 -> batchnorm

    (feat-in)-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------(feat-out)    
       256 ----------> 64------------------------>-----------------------------64------------------------>-----------------------------256-------------------------------------------->  256    

## resnet fully connected chain

In [ ]:
from IPython.display import Image
Image(filename='imgs/resnet50_v1_block_fc_mxnet.png') 

## resnet Network

  - Creating network and blocks using monk from scratch will be dealt in different roadmap series

In [ ]:
from IPython.display import Image
Image(filename='imgs/resnet50_v1_mxnet.png') 

## Properties

  - This network 
    - has 12 type-1 blocks
    - has 4 type-2 blocks
    - post these blocks the type-3 (fc) block exists


<a id='3'></a>
# Train the classifier

In [ ]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

<a id='4'></a>
# Run inference on trained classifier

In [ ]:
gtf = prototype(verbose=1);
gtf.Prototype("Project", "resnet-intro", eval_infer=True);

In [ ]:
output = gtf.Infer(img_name = "natural-images/test/test1.jpg");
from IPython.display import Image
Image(filename='natural-images/test/test1.jpg') 

In [ ]:
output = gtf.Infer(img_name = "natural-images/test/test2.jpg");
from IPython.display import Image
Image(filename='natural-images/test/test2.jpg') 

In [ ]:
output = gtf.Infer(img_name = "natural-images/test/test3.jpg");
from IPython.display import Image
Image(filename='natural-images/test/test3.jpg') 

# Goals Completed


### Train a classifier using resnet50 on natural-images dataset

### Understand what lies inside resnet50 network